### ISSUE TITLE
Improve dbListTables

### ISSUE THREAD

**Issue by ebocher** [hannesmuehleisen/renjin-dbi#10]
Tested with H2GIS and H2 database.
dbListTables is not able to list View or other exotic table as "TABLE LINK", "EXTERNAL".
It will be nice to improve that so a user will be able for example to list linked files (see http://www.h2gis.org/docs/dev/FILE_TABLE/) as it is supported by H2 database engine.

**Comment by hannes** [comment]
Not particularly surprised by this. We only ask for tables currently. We could extend https://github.com/hannesmuehleisen/renjin-dbi/blob/master/DBI/src/main/java/org/renjin/cran/DBI/JDBCUtils.java#L29 but many of these types are DB-specific. Another option is to overwrite the method for H2 alone.

**Comment by ebocher** [comment]
Ok. I'm on it
https://github.com/ebocher/renjin-dbi/tree/h2customTables

**Comment by ebocher** [comment]
Dear @hannesmuehleisen ,

I added a new method in DBI {JDBCUtils} . It's possible to set a list of table types. See https://github.com/ebocher/renjin-dbi/blob/h2customTables/DBI/src/main/java/org/renjin/cran/DBI/JDBCUtils.java#L46
In RH2GIS, https://github.com/ebocher/renjin-dbi/blob/h2customTables/RH2GIS/src/main/R/h2gis.R#L23 the functions dbListTables, dbExistTable  are overridden to use getTables(table types).
The test https://github.com/ebocher/renjin-dbi/blob/h2customTables/RH2GIS/src/test/R/h2gis-test.R#L14 works well.  However I have a problem with the others functions. eg : dbRemoveTable(con, "LINKED_SHAPE");
I'm not comfortable with R programming language. So my question is : 
Do I need to override all DBI functions in RH2GIS and RH2 or is there a simple mechanism to override only the dbListTables function ?
Advice are welcome ;-)

**Comment by hannes** [comment]
See also https://github.com/rstats-db/DBI/issues/141 for a DBI issue about the same problem. And yes, you can override only a single function. See https://github.com/hannesmuehleisen/renjin-dbi/blob/master/MonetDB.R/src/main/R/monetdb.R#L12 for an example how to do this.

**Comment by ebocher** [comment]
In the h2gis.R I try

``` R
dbListTables.h2gis <- function(con) {
    JDBCUtils$getTables(con$conn, c("TABLE", "VIEW", "TABLE LINK", "EXTERNAL")) 
}
```

but it doesn't work. If I set

``` R
dbListTables      <- function(con, ...)       UseMethod("dbListTables") 
dbListTables.JDBCConnection <- function(con) {
    JDBCUtils$getTables(con$conn, c("TABLE", "VIEW", "TABLE LINK", "EXTERNAL")) 
}
```

It works but It seems than all methods that use dbListTables must be overridden.

This as a great puzzle...

**Comment by hannes** [comment]
is `h2gis` in the `class(con)` return value for your connection?

**Comment by ebocher** [comment]
yes 

dbListTables.h2gis <- function(con)  returns the same result as the parent function dbListTables in DBI.

**Comment by hannes** [comment]
try calling it `dbListTables.H2GIS`

**Comment by ebocher** [comment]
The same !

**Comment by ebocher** [comment]
Dear @hannesmuehleisen 

I'm absolutely in the desert. I check http://adv-r.had.co.nz/Functions.html and test another approach 

In DBI.R

``` R

dbListTables.JDBCConnection <- function(con, tableTypes = c("TABLE")) {
    JDBCUtils$getTables(con$conn, tableTypes)   
}

```

In h2gis.R

``` R
dbListTables      <- function(con, tableTypes = c("TABLE", "VIEW", "TABLE LINK", "EXTERNAL"))   UseMethod("dbListTables") 



```

but it doesn't work.

**Comment by ebocher** [comment]
Finally It seems that the best way will be to copy paste the required functions in H2GIS and H2 packages.

**Comment by ebocher** [comment]
Dear @hannesmuehleisen ,

I'm working on this issue to fix it. There are two alternatives.
 1. The most easy
Add an static array in JBDCUtils

```java
  static String [] tableTypes =  new String[]{"TABLE", "VIEW", "TABLE LINK", "EXTERNAL"};
/**
     * List the tables stored in the database
     *
     * @param con
     * @return
     */
    public static StringVector getTables(Connection con) {       
        try {
            StringVector.Builder sb = new StringVector.Builder();
            DatabaseMetaData dbm = con.getMetaData();
            ResultSet rsm = dbm.getTables(con.getCatalog(), null, null, tableTypes);

            while (rsm.next()) {
                sb.add(rsm.getString("TABLE_NAME"));
            }
            rsm.close();
            return sb.build();
        } catch (SQLException e) {
            throw new EvalException(e);
        }
    }
``` 

2. Duplicate the DBI methods in H2 and H2GIS packages 

What do you think?

**Comment by hannes** [comment]
It depends, are these constants defined by the standard? If so, it seems fine. If they are db-dependent, it would be better off in the individual backends

**Comment by ebocher** [comment]
There is some standardized table types but also db-dependent type names. So I will push a proposal to offer a way to custom the table types.

**Comment by ebocher** [comment]
I have send a PR about this issue

**Comment by ebocher** [comment]
Fix with #12
